### DISCUSSION TITLE
Ensuring the same QueryClient

### DISCUSSION THREAD

**Discussion by shongs27** [pmndrs/jotai#1928] [General] [answered]
https://jotai.org/docs/integrations/query#referencing-the-same-instance-of-query-client-in-your-project

The content of the above official document is understood to mean that you should always use a **Provider** to ensure the use of **the same QueryClient**

However, even if you import and use queryClientAtom as get in the following way, you could not find any problems. `get(queryClientAtom)`

Is this the recommended method? Is there a side effect that this will bring? If not, why is this usage not in the official document?

```
import {
  atomsWithMutation,
  atomsWithQuery,
  queryClientAtom,
} from 'jotai-tanstack-query';

export const [, messageReadAtom] = atomsWithMutation<
  unknown,
  AxiosError,
  unknown
>((get) => ({
  mutationFn: ...
  onSuccess: () => {
    get(queryClientAtom).invalidateQueries([
      '/messages',
      get(headerInfo),
      get(messageStatusAtom),
      get(messageTypeAtom),
    ]);
    get(queryClientAtom).invalidateQueries([
      `/messages/${get(messageIdAtom)}`,
      get(headerInfo),
    ]);
  },
}));
```

**Comment by dai-shi** [comment, marked as answer]
We just didn't know such a usage.
It doesn't seem to have any downside from the Jotai perspective. (Except that any change of dependency atom value will trigger update for `messageReadAtom`.)
Not sure if there's any issue in terms of TanStack Query usage.

**Reply by shongs27** [reply]
Thank you. I understood. There could be various uses that are not in the official document. I'm not used to these things.

If you have time, I'd like to ask you one more question. I'm a fan watching your activities. I hope to be a developer who can understand and create helpful libraries like you. There are no video lectures in this kind of work, and I don't really understand it in depth in my actual work.

How can I customize a library with a deep understanding like you? I'm thinking about how to start. Can you give me some advice that I really need?
The operating principles obtained through blogs or official documents are always insufficient or incomprehensible. (Except for the new React official document.. that is cool) How can i move on?

**Reply by dai-shi** [reply]
I think reading source code from some OSS projects should help. Learning from multiple projects would be good to get broader knowledge.
